### PR TITLE
Fix task run count when a task is configured with `retry_delay_seconds`

### PR DESCRIPTION
**Pull request by desertaxle**
This PR moves the increment of a task run's `run_count` closer to the user code execution to ensure it's correctly counted.

Closes https://github.com/PrefectHQ/prefect/issues/15422

### CODE REVIEW

File: tests/events/client/instrumentation/test_task_run_state_change_events.py [hunk 1]
```diff
@@ -139,7 +139,7 @@ def happy_path():
             },
             "flow_run_run_count": 1,
             "name": task_run.name,
-            "run_count": 1,
+            "run_count": 0,
```
Comment (zzstoatzz) [inline tests/events/client/instrumentation/test_task_run_state_change_events.py:142]: wait isn't this test showing that we went from PENDING -> RUNNING -> COMPLETED? if so, why shouldn't this be 1?

Comment (desertaxle) [reply tests/events/client/instrumentation/test_task_run_state_change_events.py:142]: This is when it transitions from `PENDING` to `RUNNING` so the user code hasn't been run yet. When it transitions from `RUNNING` to `COMPLETED`, then the `run_count` is 1. I think we were incrementing the run count too early before.

Comment (zzstoatzz) [reply tests/events/client/instrumentation/test_task_run_state_change_events.py:142]: ahh - i see!

**Review by zzstoatzz — APPROVED** [review]
lgtm!

**Review by cicdw — COMMENTED** [review]
@desertaxle apparently my review submission didn't go through last night - want to make sure you see this comment in case it's important

File: src/prefect/task_engine.py [hunk 2]
```diff
@@ -364,7 +364,6 @@ def begin_run(self):
         new_state = Running()
 
         self.task_run.start_time = new_state.timestamp
-        self.task_run.run_count += 1
```
[the remark this answers is not in the document]
Comment (desertaxle) [reply src/prefect/task_engine.py:367]: Gotcha, I was interpreting it as the number of times the user code has been run. I'll submit a PR to revert back to the previous behavior.

[the remark this answers is not in the document]
Comment (desertaxle) [reply src/prefect/task_engine.py:367]: @cicdw Update is in https://github.com/PrefectHQ/prefect/pull/15436

**Pull request merged**